### ISSUE TITLE
Migrate to OsPath

### ISSUE THREAD

**Issue by kokobd** [haskell/haskell-language-server#3046]
We have a better alternative to FilePath now. See: https://hasufell.github.io/posts/2022-06-29-fixing-haskell-filepaths.html

What we should do:
1. Change `NormalizedFilePath` to `NormalizedOsPath` after the issue in lsp to support OsPath https://github.com/haskell/lsp/issues/445 is solved. Profiling it to make sure the performance improves.
2. Change other usages of `FilePath` to `OsPath`

However, this can not be done until `ghc-9.6` is released (ships the new `filepath` library), or all ghc shipped libraries are available on Hackage.
> Many GHC bundled libraries do not have new versions released to Hackage (for example ghci and hpc, their versions on hackage is very old). Thus, cabal solver fails if I add filepath ^>= 1.4.100.0 to a project that depends on ghc, which then depends on ghci. The installed ghci links against filepath-1.4.2.2 in this case, and cabal can not rebuild a ghci with filepath ^>= 1.4.100.0, as the source code is not found on hackage.

**Comment by pepeiborra** [comment]
> Thus, cabal solver fails if I add filepath ^>= 1.4.100.0 to a project that depends on ghc, which then depends on ghci.

Oh bummer. Well, if we really want to, it should be ok to embed `OsPath` and combinators in a `filepath-compat` package.

**Comment by hasufell** [comment]
https://gitlab.haskell.org/ghc/ghc/-/issues/21887#note_443922

It may be possible to provide updated GHC releases that include new filepath. But this is still in discussion.

**Comment by kokobd** [comment]
I don't know why, but the overall memory usage decreased from 1.5G to 700M
The data is collected by launching `haskell-language-server` from command line with profiling RTS options enabled.
<img width="1137" alt="image" src="https://user-images.githubusercontent.com/16440269/181737382-e6057143-f387-4a34-8260-a8c448d4944a.png">
<img width="1137" alt="image" src="https://user-images.githubusercontent.com/16440269/181737424-bf63f673-514b-4af8-a6ce-7bbee848b654.png">

**Comment by hasufell** [comment]
@kokobd you mean between String and OsPath?

**Comment by kokobd** [comment]
> @kokobd you mean between String and OsPath?

@hasufell Yes. But when loading HLS executable into VSCode, I no longer witness such a big difference. I'm still investigating the reason.

**Comment by pepeiborra** [comment]
Can you run cabal bench ghcide and share the contents of the ghcide/bench-results folder minus the binaries ?

**Comment by kokobd** [comment]
`cabal bench ghcide` is failing with master & ghc 9.2.3. Verified by creating a fresh new Gitpod workspace from master.
<img width="675" alt="image" src="https://user-images.githubusercontent.com/16440269/181904655-08fe12c9-0c3d-44ba-961c-fba5267844f7.png">

**Comment by pepeiborra** [comment]
I'll take a look, but in the meantime CI shows that it works fine on 8.10.7. Would you be able to run it there?

**Comment by pepeiborra** [comment]
Ran the edit benchmark on Cabal 3.6.3.0, results seem quite promising:
![edit](https://user-images.githubusercontent.com/26626/181925830-59666bfe-f7b1-43e2-834a-eb7b56622f8e.svg)

**Comment by kokobd** [comment]
Next steps:
1. Upgrade `lsp` and `lsp-types` in HLS to 1.5. It has some breaking changes.
2. Either extend [`filepath-compat`](https://github.com/kokobd/filepath-compat) to support older versions of GHC (currently it only supports ghc 9.2.3), or release GHC minor versions with new `filepath` package. Support of some old GHC might need to be dropped. See: https://gitlab.haskell.org/ghc/ghc/-/issues/21887

**Comment by kokobd** [comment]
Real `OsPath` will only be available since GHC 9.6. Instead, I use `ShortByteString` within `NormalizedFilePath`. The result is still very promising.

Soon, we will benefit from the performance improvement in all versions of GHC supported by HLS.

<img width="797" alt="image" src="https://user-images.githubusercontent.com/16440269/186920992-bc5f5db9-89f5-4401-8a0e-b8f9fe535ea3.png">
<img width="803" alt="image" src="https://user-images.githubusercontent.com/16440269/186922533-e5f184da-8a35-40a5-83bf-b05d99ba1644.png">

Next steps:
1. Merge https://github.com/haskell/lsp/pull/446 and #3067. (almost done)
2. Eliminate the usage of partial functions introduced by this PR gradually.

**Comment by michaelpj** [comment]
I think this is mostly done. Was there more to do?

**Comment by hasufell** [comment]
I haven't looked at it, but I wonder how you do file operations etc. I'm guessing you convert back to filepath somewhere? In that case the conversion is not complete

**Comment by michaelpj** [comment]
It looks like the situation is:
- `lsp` is set up for `OsPath`
- HLS hasn't migrated at all, and in the interests of keeping it simple we probably won't until 9.6 is the oldest version we support, so we don't need conditional logic and can just use `OsPath` everywhere

**Comment by hasufell** [comment]
So in ten years, I guess?

**Comment by michaelpj** [comment]
Two major GHC versions, so yeah, a while. Or I guess if someone is keen and wants to do it with conditional logic they could do!

**Comment by crtschin** [comment]
It's time: https://github.com/haskell/lsp/issues/608.
